src/predictor/gpu_predictor.cu

```diff
@@ -324,22 +336,31 @@ class GPUPredictor : public xgboost::Predictor {
                 h_nodes.begin() + h_tree_segments[tree_idx - tree_begin]);
     }
 
-    size_t i_batch = 0;
     size_t batch_offset = 0;
     for (auto &batch : dmat->GetRowBatches()) {
-      CHECK(i_batch == 0 || devices_.Size() == 1) << "External memory not supported for multi-GPU";
-      // out_preds have been sharded and resized in InitOutPredictions()
-      batch.offset.Shard(GPUDistribution::Overlap(devices_, 1));
+      GPUSet devices = GPUSet::All(param_.gpu_id, param_.n_gpus, batch.Size());
+      ConfigureShards(devices);
+      std::vector<size_t> out_preds_offsets;
+      PredictionDeviceOffsets(out_preds->Size(), batch_offset, batch.Size(),
+                              model.param.num_output_group, &out_preds_offsets);
+      out_preds->Reshard(GPUDistribution::Explicit(devices_, out_preds_offsets));
+
+      batch.offset.Reshard(GPUDistribution::Overlap(devices_, 1));
       std::vector<size_t> device_offsets;
       DeviceOffsets(batch.offset, &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
+
       dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
-        shard.PredictInternal(batch, dmat->Info(), out_preds, batch_offset, model,
+        shard.PredictInternal(batch, dmat->Info(), out_preds, model,
                               h_tree_segments, h_nodes, tree_begin, tree_end);
       });
       batch_offset += batch.Size() * model.param.num_output_group;
-      i_batch++;
     }
+
```

Comment:
how about extracting the existing distribution before iterating over the batches and reinstating it after?

before looping:
```
const GPUDistribution dist = out_preds->Distribution();
```

after looping:
```
out_preds->Reshard(dist);
```

this way, this method need not worry about how it was sharded before.

Reply:
Not sure if we need to do this, as we used to shard `out_preds` ourselves in `InitOutPredictions()`.

Reply:
the existing code may have assumed this behavior (or explicitly shard with with similar configuration, which may break) - i.e. the resultant out_preds has been sharded based on the number of rows and number of classes. although, this specific change of extracting and reinstating the distribution may no longer be required, as you removed the sharding from `InitOutPredictions`, it should be required to preserve current behavior; else, clients may not be able to use this unless they reshard it based on previous assumptions.

Reply:
Right, the final resharding after all the batches are done should preserve current behavior.